File: lib/whatsapp.ts
// Unified WhatsApp sender. Switch provider via WA_PROVIDER env.
// Provider: meta     — Meta WhatsApp Cloud API (free 1000/mo)
// Provider: local    — whatsapp-web.js node bridge (Puppeteer-based)
// Provider: evolution — Evolution API (Baileys engine, no Puppeteer, recommended)

type SendResult = { ok: boolean; error?: string; raw?: any; simulated?: boolean; message?: string };

export async function sendWhatsApp(toRaw: string, body: string, documentBuffer?: ArrayBuffer | null): Promise<SendResult> {
  const to = toRaw.replace(/[^\d]/g, "");
  const provider = (process.env.WA_PROVIDER || "none").trim().toLowerCase();

  console.log(`[WhatsApp] Provider: ${provider} | To: ${to}`);
  console.log(`[WhatsApp] Message: ${body}`);
  if (documentBuffer) {
    console.log(`[WhatsApp] Document attached (${documentBuffer.byteLength} bytes)`);
  }

  if (provider === "none" || provider === "demo") {
    return {
      ok: true,
      simulated: true,
      message: "WhatsApp is in demo mode — message logged but not sent. Set WA_PROVIDER=evolution to enable real WhatsApp."
    };
  }

  if (provider === "meta")      return sendViaMeta(to, body, documentBuffer);
  if (provider === "local")     return sendViaLocal(to, body, documentBuffer);
  if (provider === "evolution") return sendViaEvolution(to, body, documentBuffer);

  return { ok: false, error: `Unknown WA_PROVIDER: ${provider}` };
}

// ─── Meta (WhatsApp Business Cloud API) ───────────────────────────────────────

async function sendViaMeta(to: string, body: string, documentBuffer?: ArrayBuffer | null): Promise<SendResult> {
  const token = process.env.META_WA_TOKEN;
  const phoneId = process.env.META_WA_PHONE_ID;
  if (!token || !phoneId) return { ok: false, error: "Meta WA env missing (META_WA_TOKEN / META_WA_PHONE_ID)" };

  try {
    if (documentBuffer) {
      const formData = new FormData();
      const blob = new Blob([documentBuffer], { type: "application/pdf" });
      formData.append("messaging_product", "whatsapp");
      formData.append("to", to);
      formData.append("type", "document");
      formData.append("document", blob, "document.pdf");
      formData.append("caption", body);

      const res = await fetch(`https://graph.facebook.com/v21.0/${phoneId}/messages`, {
        method: "POST",
        headers: { Authorization: `Bearer ${token}` },
        body: formData
      });
      const json = await res.json();
      if (!res.ok) return { ok: false, error: JSON.stringify(json), raw: json };
      return { ok: true, raw: json };
    }

    const res = await fetch(`https://graph.facebook.com/v21.0/${phoneId}/messages`, {
      method: "POST",
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
      body: JSON.stringify({ messaging_product: "whatsapp", to, type: "text", text: { body } })
    });
    const json = await res.json();
    if (!res.ok) return { ok: false, error: JSON.stringify(json), raw: json };
    return { ok: true, raw: json };
  } catch (e: any) {
    return { ok: false, error: e.message };
  }
}

// ─── Local (whatsapp-web.js bridge, legacy) ────────────────────────────────────

async function sendViaLocal(to: string, body: string, documentBuffer?: ArrayBuffer | null): Promise<SendResult> {
  let url = process.env.LOCAL_WA_URL;
  const secret = process.env.LOCAL_WA_SECRET || process.env.SECRET;
  if (!url) return { ok: false, error: "LOCAL_WA_URL missing" };
  if (url.endsWith("/")) url = url.slice(0, -1);

  try {
    let media = null;
    if (documentBuffer) {
      const base64 = Buffer.from(documentBuffer).toString("base64");
      media = { mimetype: "application/pdf", filename: "Invoice.pdf", data: base64 };
    }

    const res = await fetch(`${url}/send`, {
      method: "POST",
      headers: { "Content-Type": "application/json", "x-secret": secret || "" },
      body: JSON.stringify({ to, body, media })
    });
    const json = await res.json().catch(() => ({}));
    if (!res.ok) return { ok: false, error: JSON.stringify(json) };
    return { ok: true, raw: json };
  } catch (e: any) {
    return { ok: false, error: e.message };
  }
}

// ─── Evolution API (recommended — Baileys engine, no Puppeteer) ───────────────
//
// Setup:
//   EVOLUTION_API_URL      = http://<your-ec2-ip>:8080
//   EVOLUTION_API_KEY      = your-api-key
//   EVOLUTION_INSTANCE     = gymapp   (instance name you created)
//
// Quick start on EC2:
//   docker run -d --name evolution -p 8080:8080 \
//     -e AUTHENTICATION_TYPE=apikey \
//     -e AUTHENTICATION_API_KEY=your-api-key \
//     -v evolution_data:/evolution/instances \
//     atendai/evolution-api:latest
//
// Then POST http://<ip>:8080/instance/create  { "instanceName":"gymapp","qrcode":true }
// Open http://<ip>:8080/instance/connect/gymapp → scan QR

async function sendViaEvolution(to: string, body: string, documentBuffer?: ArrayBuffer | null): Promise<SendResult> {
  const baseUrl = process.env.EVOLUTION_API_URL;
  const apiKey  = process.env.EVOLUTION_API_KEY;
  const instance = process.env.EVOLUTION_INSTANCE || "gymapp";

  if (!baseUrl) return { ok: false, error: "EVOLUTION_API_URL missing" };
  if (!apiKey)  return { ok: false, error: "EVOLUTION_API_KEY missing" };

  // Evolution API expects number in format: 919876543210@s.whatsapp.net
  const phone = `${to}@s.whatsapp.net`;
  const url = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
  const headers = { "Content-Type": "application/json", apikey: apiKey };

  try {
    // ── Send PDF / document ──────────────────────────────────────────────────
    if (documentBuffer) {
      const base64 = Buffer.from(documentBuffer).toString("base64");
      const res = await fetch(`${url}/message/sendMedia/${instance}`, {
        method: "POST",
        headers,
        body: JSON.stringify({
          number: phone,
          mediatype: "document",
          mimetype: "application/pdf",
          caption: body,
          media: base64,
          fileName: "Invoice.pdf"
        })
      });
      const json = await res.json().catch(() => ({}));
      if (!res.ok) return { ok: false, error: JSON.stringify(json), raw: json };
      return { ok: true, raw: json };
    }

    // ── Send plain text ──────────────────────────────────────────────────────
    const res = await fetch(`${url}/message/sendText/${instance}`, {
      method: "POST",
      headers,
      body: JSON.stringify({ number: phone, text: body })
    });
    const json = await res.json().catch(() => ({}));
    if (!res.ok) return { ok: false, error: JSON.stringify(json), raw: json };
    return { ok: true, raw: json };
  } catch (e: any) {
    return { ok: false, error: e.message };
  }
}

File: backend/api/whatsapp.py
import os
import re
import base64
import requests

def send_whatsapp(to_raw: str, body: str, document_bytes: bytes = None) -> dict:
    """
    Unified WhatsApp sender. Selects provider based on WA_PROVIDER env variable.
    """
    to = re.sub(r'\D', '', to_raw)
    provider = os.getenv("WA_PROVIDER", "none").strip().lower()

    print(f"[WhatsApp] Provider: {provider} | To: {to}")
    print(f"[WhatsApp] Message: {body}")
    if document_bytes:
        print(f"[WhatsApp] Document attached ({len(document_bytes)} bytes)")

    if provider in ("none", "demo", ""):
        return {
            "ok": True,
            "simulated": True,
            "message": "WhatsApp is in demo mode — message logged but not sent. Set WA_PROVIDER=evolution to enable real WhatsApp."
        }

    if provider == "meta":
        return send_via_meta(to, body, document_bytes)
    if provider == "local":
        return send_via_local(to, body, document_bytes)
    if provider == "evolution":
        return send_via_evolution(to, body, document_bytes)

    return {"ok": False, "error": f"Unknown WA_PROVIDER: {provider}"}

def send_via_meta(to: str, body: str, document_bytes: bytes = None) -> dict:
    token = os.getenv("META_WA_TOKEN")
    phone_id = os.getenv("META_WA_PHONE_ID")
    if not token or not phone_id:
        return {"ok": False, "error": "Meta WA env missing (META_WA_TOKEN / META_WA_PHONE_ID)"}

    url = f"https://graph.facebook.com/v21.0/{phone_id}/messages"
    headers = {"Authorization": f"Bearer {token}"}

    try:
        if document_bytes:
            files = {
                "document": ("document.pdf", document_bytes, "application/pdf")
            }
            data = {
                "messaging_product": "whatsapp",
                "to": to,
                "type": "document",
                "caption": body
            }
            res = requests.post(url, headers=headers, data=data, files=files)
        else:
            headers["Content-Type"] = "application/json"
            payload = {
                "messaging_product": "whatsapp",
                "to": to,
                "type": "text",
                "text": {"body": body}
            }
            res = requests.post(url, headers=headers, json=payload)

        json_res = res.json()
        if res.status_code != 200:
            return {"ok": False, "error": res.text, "raw": json_res}
        return {"ok": True, "raw": json_res}
    except Exception as e:
        return {"ok": False, "error": str(e)}

def send_via_local(to: str, body: str, document_bytes: bytes = None) -> dict:
    url = os.getenv("LOCAL_WA_URL")
    secret = os.getenv("LOCAL_WA_SECRET") or os.getenv("SECRET")
    if not url:
        return {"ok": False, "error": "LOCAL_WA_URL missing"}
    
    url = url.rstrip("/")
    headers = {"Content-Type": "application/json", "x-secret": secret or ""}
    
    try:
        media = None
        if document_bytes:
            b64_str = base64.b64encode(document_bytes).decode("utf-8")
            media = {
                "mimetype": "application/pdf",
                "filename": "Invoice.pdf",
                "data": b64_str
            }
        
        payload = {"to": to, "body": body}
        if media:
            payload["media"] = media
            
        res = requests.post(f"{url}/send", headers=headers, json=payload)
        json_res = res.json()
        if res.status_code != 200:
            return {"ok": False, "error": res.text}
        return {"ok": True, "raw": json_res}
    except Exception as e:
        return {"ok": False, "error": str(e)}

def send_via_evolution(to: str, body: str, document_bytes: bytes = None) -> dict:
    base_url = os.getenv("EVOLUTION_API_URL")
    api_key = os.getenv("EVOLUTION_API_KEY")
    instance = os.getenv("EVOLUTION_INSTANCE", "gymapp")

    if not base_url:
        return {"ok": False, "error": "EVOLUTION_API_URL missing"}
    if not api_key:
        return {"ok": False, "error": "EVOLUTION_API_KEY missing"}

    phone = f"{to}@s.whatsapp.net"
    url = base_url.rstrip("/")
    headers = {"Content-Type": "application/json", "apikey": api_key}

    try:
        if document_bytes:
            b64_str = base64.b64encode(document_bytes).decode("utf-8")
            payload = {
                "number": phone,
                "mediatype": "document",
                "mimetype": "application/pdf",
                "caption": body,
                "media": b64_str,
                "fileName": "Invoice.pdf"
            }
            res = requests.post(f"{url}/message/sendMedia/{instance}", headers=headers, json=payload)
        else:
            payload = {
                "number": phone,
                "text": body
            }
            res = requests.post(f"{url}/message/sendText/{instance}", headers=headers, json=payload)

        json_res = res.json()
        if res.status_code not in (200, 201):
            return {"ok": False, "error": res.text, "raw": json_res}
        return {"ok": True, "raw": json_res}
    except Exception as e:
        return {"ok": False, "error": str(e)}

def get_whatsapp_status() -> dict:
    """
    Checks connection status of the WhatsApp instance.
    """
    provider = os.getenv("WA_PROVIDER", "none").strip().lower()
    if provider in ("none", "demo", ""):
        return {"ok": True, "provider": "demo", "status": "open", "message": "Demo mode"}
    
    if provider == "evolution":
        base_url = os.getenv("EVOLUTION_API_URL")
        api_key = os.getenv("EVOLUTION_API_KEY")
        instance = os.getenv("EVOLUTION_INSTANCE", "gymapp")
        
        if not base_url or not api_key:
            return {"ok": False, "error": "Evolution credentials missing"}
            
        url = f"{base_url.rstrip('/')}/instance/connectionState/{instance}"
        headers = {"apikey": api_key}
        try:
            res = requests.get(url, headers=headers, timeout=10)
            if res.status_code == 200:
                data = res.json()
                state = data.get("instance", {}).get("state", "close")
                return {"ok": True, "provider": "evolution", "status": state}
            return {"ok": False, "error": f"Evolution status error: {res.text}"}
        except Exception as e:
            return {"ok": False, "error": str(e)}
            
    return {"ok": True, "provider": provider, "status": "unknown"}

def get_whatsapp_qr() -> dict:
    """
    Fetches the connection QR code for WhatsApp.
    """
    provider = os.getenv("WA_PROVIDER", "none").strip().lower()
    if provider in ("none", "demo", ""):
        return {"ok": True, "provider": "demo", "message": "Demo mode - no QR code required"}
        
    if provider == "evolution":
        base_url = os.getenv("EVOLUTION_API_URL")
        api_key = os.getenv("EVOLUTION_API_KEY")
        instance = os.getenv("EVOLUTION_INSTANCE", "gymapp")
        
        if not base_url or not api_key:
            return {"ok": False, "error": "Evolution credentials missing"}
            
        url = f"{base_url.rstrip('/')}/instance/connect/{instance}"
        headers = {"apikey": api_key}
        try:
            res = requests.get(url, headers=headers, timeout=15)
            if res.status_code == 200:
                data = res.json()
                # Return code and base64
                return {
                    "ok": True, 
                    "provider": "evolution", 
                    "code": data.get("code"),
                    "base64": data.get("base64")
                }
            return {"ok": False, "error": f"Evolution connect error: {res.text}"}
        except Exception as e:
            return {"ok": False, "error": str(e)}
            
    return {"ok": False, "error": f"QR code generation not supported for provider: {provider}"}

